docs/tools.md

```diff
@@ -720,15 +723,556 @@ def my_flaky_tool(query: str) -> str:
+result = agent.run_sync('Call all the tools')
+print(LOG)
+"""
+[
+    "Calling tool 'temperature_celsius' with args: {'city': 'a'}",
+    "Calling tool 'temperature_fahrenheit' with args: {'city': 'a'}",
+    "Calling tool 'weather_conditions' with args: {'city': 'a'}",
+    "Calling tool 'current_time' with args: {}",
+    "Finished calling tool 'temperature_celsius' with result: 21.0",
+    "Finished calling tool 'temperature_fahrenheit' with result: 69.8",
+    'Finished calling tool \'weather_conditions\' with result: "It\'s raining"',
+    "Finished calling tool 'current_time' with result: datetime.datetime(...)",
+]
+"""
+```
+
+1. We use [`TestModel`][pydantic_ai.models.test.TestModel] here as it will automatically call each tool.
+
+_(This example is complete, it can be run "as is")_
+
+##### Modifying Toolsets During a Run
+
+You can change the `WrapperAgent`'s `wrapped` property during an agent run to swap out one toolset for another starting at the next run step.
+
+To add or remove available toolsets, you can wrap a [`CombinedToolset`](#combining-toolsets) and replace it during the run with one that can include fewer, more, or entirely different toolsets.
+
+```python {title="wrapper_toolset.py" requires="function_toolset.py"}
+from function_toolset import weather_toolset, datetime_toolset
+
+from pydantic_ai import Agent, RunContext
+from pydantic_ai.models.test import TestModel
+from pydantic_ai.toolsets import WrapperToolset, FunctionToolset
+
+togglable_toolset = WrapperToolset(weather_toolset)
+
+def toggle(ctx: RunContext[WrapperToolset]):
+    if ctx.deps.wrapped == weather_toolset:
+        ctx.deps.wrapped = datetime_toolset
+    else:
+        ctx.deps.wrapped = weather_toolset
+
+test_model = TestModel() # (1)!
+agent = Agent(
+    test_model,
+    deps_type=WrapperToolset, # (2)!
+    toolsets=[togglable_toolset, FunctionToolset([toggle])]
+)
+print([t.name for ts in agent.toolsets for t in ts.tool_defs])
+#> ['temperature_celsius', 'temperature_fahrenheit', 'conditions', 'toggle']
+
+result = agent.run_sync('Toggle the toolset', deps=togglable_toolset)
+print([t.name for t in test_model.last_model_request_parameters.function_tools])
+#> ['now', 'toggle']
+```
+
+1. We're using [`TestModel`][pydantic_ai.models.test.TestModel] here because it makes it easy to see which tools were available on each run.
+2. We're using the agent's dependencies to give the `toggle` tool access to the `togglable_toolset` via the `RunContext` argument
+
+### Building a Custom Toolset
+
+To define a fully custom toolset with its own logic to list available tools and handle them being called, you can subclass [`BaseToolset`][pydantic_ai.toolsets.BaseToolset] or [`AsyncBaseToolset`][pydantic_ai.toolsets.AsyncBaseToolset], depending on whether listing the available tools can be done synchronously or requires an asynchronous network request. Tool calls themselves are always implemented asynchronously.
```

Comment:
At some point we might want to support synchronous tools, e.g. for computationally intensive tools. I guess we don't need to start there but probably shouldn't commit to APIs we'll regret. I suspect it's fine just pointing it out.